### PR TITLE
Formalize transfer state machines

### PR DESCRIPTION
**Pull request by jophish**
Previously, the progression of transfer states was underspecified, as was the exact process for completing a transfer out, since transfers out require interaction from the end-user in order to complete. This PR adds in several, more descriptive transfer states, and specifies formal state machines for both transfers in and out.

I recommend checking the PR out in `View File` mode so you can see the rendered `mermaid` state machines.

### CODE REVIEW

File: fiatconnect-api.md [hunk 1]
```diff
@@ -1710,14 +1713,78 @@ This endpoint MUST fail when the user has no transfer on file with the provided
...
+from the user's fiat account but before crypto funds are sent to the user, the server MUST return any fiat funds to the user's fiat account in case
+of such a failure.
+
+```mermaid
+graph LR
+   TransferStarted---->TransferPending
+   TransferPending---->TransferFiatFundsDebited
+   TransferFiatFundsDebited---->TransferWaitingForFiatFunds
+   TransferWaitingForFiatFunds---->TransferReceivedFiatFunds
+   TransferReceivedFiatFunds---->TransferSendingCryptoFunds
```
Comment (cajubelt) [inline fiatconnect-api.md:1736]: as a provider, would I ever want to do anything between `TransferReceivedFiatFunds` and `TransferSendingCryptoFunds`? (Could they be one state?)

File: fiatconnect-api.md [hunk 2]
```diff
@@ -1710,14 +1713,78 @@ This endpoint MUST fail when the user has no transfer on file with the provided
...
+
+Statuses for transfers in must progress through the state machine shown below, ultimately ending at one of the two *terminal* states:
+`TransferFailed` or `TransferComplete`. A server SHOULD make a best effort to reach a terminal transfer state within the `settlementTimeUpperBound`
+specified for the quote associated with the transfer, if one exists. Note that since a transfer in may fail after the server has received fiat funds
+from the user's fiat account but before crypto funds are sent to the user, the server MUST return any fiat funds to the user's fiat account in case
+of such a failure.
+
+```mermaid
+graph LR
+   TransferStarted---->TransferPending
```
Comment (cajubelt) [inline fiatconnect-api.md:1732]: why have a TransferPending state? like what happens between TransferStarted and TransferPending?

File: fiatconnect-api.md [hunk 3]
```diff
@@ -1710,14 +1713,78 @@ This endpoint MUST fail when the user has no transfer on file with the provided
...
+their fiat account with fiat funds. In case of such a failure, the server MUST return any crypto funds sent by the user back to the user's
+crypto address.
+
+```mermaid
+graph LR
+   TransferStarted---->TransferPending
+   TransferPending---->TransferReadyForUserToSendCryptoFunds
+   TransferReadyForUserToSendCryptoFunds---->TransferReceivedCryptoFunds
+   TransferReceivedCryptoFunds---->TransferFiatFundsCredited
+   TransferFiatFundsCredited---->TransferComplete
```
Comment (cajubelt) [inline fiatconnect-api.md:1770]: could these be one state? (TransferFiatFundsCredited and TransferComplete)

I'm noticing the transfer can't fail after TransferFiatFundsCredited...

File: fiatconnect-api.md [hunk 4]
```diff
@@ -1710,14 +1713,78 @@ This endpoint MUST fail when the user has no transfer on file with the provided
...
+
+   TransferStarted---->TransferFailed
+   TransferPending---->TransferFailed
+   TransferPending---->TransferAmlFailed
+   TransferAmlFailed---->TransferFailed
```
Comment (cajubelt) [inline fiatconnect-api.md:1775]: should TransferAmlFailed be a terminal state? or should TransferFailed have a failure reason? seems weird to have TransferAmlFailed as an intermediate state just to convey the reason for the transfer failure

File: fiatconnect-api.md [hunk 5]
```diff
@@ -1710,14 +1713,78 @@ This endpoint MUST fail when the user has no transfer on file with the provided
...
+
+Statuses for transfers out must progress through the state machine shown below, ultimately ending at one of the two *terminal* states:
+`TransferFailed` or `TransferComplete`. Note that at the beginning of a transfer out, a server may perform AML checks. If these AML checks
+fail, the server SHOULD progress the state machine to `TransferAmlFailed` before moving to the terminal `TransferFailed` status. A server SHOULD
+make a best effort to reach a terminal transfer state within the `settlementTimeUpperBound` specified for the quote associated with the transfer,
+if one exists.
+
+Transfers out are fundamentally different from transfer in in that they require user interaction partway through the state machine. Once a
+transfer out's state moves to `TransferReadyForUserToSendCryptoFunds`, the server should wait for the user to send the proper amount of crypto
+funds to the address specified for the transfer. Once the user has sent the proper amount of funds, the server SHOULD progress the state to
```
Comment (cajubelt) [inline fiatconnect-api.md:1757]: why "should" here? doesn't that mean they don't *have* to?

File: fiatconnect-api.md [hunk 6]
```diff
@@ -1710,14 +1713,78 @@ This endpoint MUST fail when the user has no transfer on file with the provided
...
+## 4.2. Transfers Out
+
+Statuses for transfers out must progress through the state machine shown below, ultimately ending at one of the two *terminal* states:
+`TransferFailed` or `TransferComplete`. Note that at the beginning of a transfer out, a server may perform AML checks. If these AML checks
+fail, the server SHOULD progress the state machine to `TransferAmlFailed` before moving to the terminal `TransferFailed` status. A server SHOULD
+make a best effort to reach a terminal transfer state within the `settlementTimeUpperBound` specified for the quote associated with the transfer,
+if one exists.
+
+Transfers out are fundamentally different from transfer in in that they require user interaction partway through the state machine. Once a
+transfer out's state moves to `TransferReadyForUserToSendCryptoFunds`, the server should wait for the user to send the proper amount of crypto
```
Comment (cajubelt) [inline fiatconnect-api.md:1756]: wait for how long?

Comment (cajubelt) [reply fiatconnect-api.md:1756]: at least until the quote expires, maybe?

**Review by cajubelt — COMMENTED** [review]
cool diagrams! good idea to specify this more

File: fiatconnect-api.md (hunk 1 above)
Comment (jophish) [reply fiatconnect-api.md:1736]: Yeah, they could probably safely be merged... I wanted to have clear events for receipt of fiat funds and transfer of crypto funds, but realistically, these two things will happen back to back.

File: fiatconnect-api.md (hunk 2 above)
Comment (jophish) [reply fiatconnect-api.md:1732]: Yeah this was left over from the previous design, but should probably be removed now that there's more specificity.

File: fiatconnect-api.md (hunk 6 above)
Comment (jophish) [reply fiatconnect-api.md:1756]: yeah, i like waiting until at least the quote expiration here.

**Review by jh2oman — APPROVED** [review]

File: fiatconnect-api.md [hunk 7]
```diff
@@ -1710,14 +1713,71 @@ This endpoint MUST fail when the user has no transfer on file with the provided
...
+Statuses for transfers in must progress through the state machine shown below, ultimately ending at one of the two *terminal* states:
+`TransferFailed` or `TransferComplete`. A server SHOULD make a best effort to reach a terminal transfer state within the `settlementTimeUpperBound`
+specified for the quote associated with the transfer, if one exists. Note that since a transfer in may fail after the server has received fiat funds
+from the user's fiat account but before crypto funds are sent to the user, the server MUST return any fiat funds to the user's fiat account in case
+of such a failure.
+
+```mermaid
+graph LR
+   TransferStarted---->TransferFiatFundsDebited
+   TransferFiatFundsDebited---->TransferWaitingForFiatFunds
```
Comment (cajubelt) [inline fiatconnect-api.md:1733]: sorry one more state thing-- why are TransferFiatFundsDebited and TransferWAitingForFiatFunds different states? In other words, what would a provider do in the "debited" state before it wants to enter the "waiting for fiat funds" state?

Comment (jophish) [reply fiatconnect-api.md:1733]: That's a good point. Probably nothing; I'll get rid of that state.

**Review by cajubelt — APPROVED** [review]
🥳

**Pull request merged**